src/extract-links/types.ts

```diff
@@ -0,0 +1,191 @@
+import { FrameElement } from '../types/util'
+import { Resource, DomResource } from '../types/resource' // TODO Remove need for this (recursive) import
+import { SubresourceType } from './url-attributes/types'
+
+// This alias is used to explicitly state which strings are guaranteed/presumed to be absolute URLs.
+export type UrlString = string
+
+export type Link = NonSubresourceLink | SubresourceLink
+
+interface Link_base {
+    // The link's target URL. This is the exact value as it appears in the document, and may thus be
+    // a relative URL. This property can be written to, which will modify the document.
+    target: string;
+
+    // The link's target URL as an absolute URL. This takes into account factors like the <base
+    // href="..."> tag, so usually you may prefer to use `absoluteTarget` rather than `target`.
+    // If the target is not a valid (relative) URL, absoluteTarget equals undefined.
+    readonly absoluteTarget?: UrlString;
+
+    // A boolean indicating whether the resource being linked to is normally considered a
+    // subresource of the document. For example, the `src` of an `<img>` tag specifies a subresource
+    // because the image is considered *part of* the document, while the `href` of an `<a>` or the
+    // `action` of a `<form>` merely *point to* another resource.
+    readonly isSubresource: boolean;
+
+    // Information needed to find the link in the DOM or stylesheet, for scenarios where one needs
+    // to do more than just reading or modifying the link target.
+    readonly from: Anchor;
+}
+
+// Links that refer to resources without making them ‘part of’ the document.
+// Note the WHATWG HTML spec calls such (and *only* such) links ‘hyperlinks’ <https://html.spec.whatwg.org/multipage/links.html#hyperlink>
+export interface NonSubresourceLink extends Link_base {
+    readonly isSubresource: false;
+    readonly subresourceType: undefined;
+}
+
+// Links that make their targets ‘part of’ the document.
+export interface SubresourceLink extends Link_base {
+    readonly isSubresource: true;
+
+    // Indicates the type of resource (`image`, `style`, ...). This corresponds to what is now
+    // called the 'destination' in the WHATWG fetch spec. See <https://fetch.spec.whatwg.org/#concept-request-destination>
+    readonly subresourceType?: SubresourceType;
+
+    // If the subresource is available, it can be assigned to this attribute.
+    // TODO Remove this ‘upward’ dependency.
+    resource?: Resource;
```

Comment:
Until now, a Resource contains Links and a Link can contain a Resource, to create their dependency tree structure. I am thinking about how (or whether) to make links not know anything about Resources, especially with an eye on factoring out `src/extract-links` into its own module. I suppose this will be addressed while modifying the crawling process.